### PR TITLE
Random crash fix - preemptible workitems

### LINKED ISSUE #1038: Scheduled CI/CD run hit crash in unit test
https://github.com/microsoft/ebpf-for-windows/runs/6232634011?check_suite_focus=true

### PR DESCRIPTION
**Pull request by saxena-anurag**
## Description

**Issue**
native module in ebpfcore queues preemptible workitems in some scenarios. These workitems access the native module hash table which can be freed before the workitem is executed (in service stop scenario). This race condition is causing random crashes in the tests.

**Fix** 
The workitem routine does not really need to access the hash table, as it is only to unload the native driver. Modified the code now to ensure workitem routine does not touch any data structures, and only calls _unload driver_ function.

There is already code present to ensure the service does not stop until all the workitems have completed. So that aspect is already covered.

## Testing

Existing tests cover this.

## Documentation

NA

Fixes #1038

### CODE REVIEW

File: libs/execution_context/ebpf_native.c [hunk 1]
```diff
@@ -1349,13 +1342,26 @@ ebpf_native_unload(_In_ const GUID* module_id)
 
     memcpy(service_name, module->service_name, service_name_length);
 
+    // Create a workitem if we are running at DISPATCH
```
Comment (dthaler) [inline libs/execution_context/ebpf_native.c:1345]:
```suggestion
    // Create a work item if we are running at DISPATCH.
```

File: libs/execution_context/ebpf_native.c [hunk 2]
```diff
@@ -1306,7 +1300,8 @@ ebpf_native_unload(_In_ const GUID* module_id)
     ebpf_native_module_t* module = NULL;
     wchar_t* service_name = NULL;
     size_t service_name_length;
-    bool unload_module = false;
+    bool queue_workitem = false;
```
Comment (dthaler) [inline libs/execution_context/ebpf_native.c:1303]:
```suggestion
    bool queue_work_item = false;
```
nit: consistency with line 1304

File: libs/execution_context/ebpf_native.c [hunk 3]
```diff
@@ -1349,13 +1342,26 @@ ebpf_native_unload(_In_ const GUID* module_id)
 
     memcpy(service_name, module->service_name, service_name_length);
 
+    // Create a workitem if we are running at DISPATCH
+    if (!ebpf_is_preemptible()) {
+        result = ebpf_allocate_preemptible_work_item(&work_item, _ebpf_native_unload_workitem, service_name);
```
Comment (dthaler) [inline libs/execution_context/ebpf_native.c:1347]:
```suggestion
        result = ebpf_allocate_preemptible_work_item(&work_item, _ebpf_native_unload_work_item, service_name);
```

File: libs/platform/user/ebpf_platform_user.cpp [hunk 4]
```diff
@@ -510,7 +512,7 @@ ebpf_is_non_preemptible_work_item_supported()
 bool
 ebpf_is_preemptible_work_item_supported()
 {
-    return false;
+    return true;
 }
```
Comment (dthaler) [inline libs/platform/user/ebpf_platform_user.cpp:516]: Can't this function be deleted now?  As far as I can find, the only caller was the ebpf_native.c call that you deleted in this PR.  Same in ebpf_platform_kernel.c

Comment (saxena-anurag) [reply libs/platform/user/ebpf_platform_user.cpp]: Removed.

**Review by dthaler — DISMISSED** [review]

**Review by dthaler — APPROVED** [review]

**Review by Alan-Jowett — APPROVED** [review]

**Pull request merged**